content/articles/api-authentication-with-django-knox-and-postman-testing/index.md

```diff
@@ -0,0 +1,343 @@
+
+
+class LoginSerializer(serializers.Serializer):
+    username = serializers.CharField()
+    password = serializers.CharField()
+
+    def validate(self, data):
+        user = authenticate(**data)
+        if user and user.is_active:
+            return user
+        raise serializers.ValidationError('Incorrect Credentials Passed.')
+```
+
+The `UserSerializer` class maps the `User` model and extracts the mentioned fields for the serialization.
+
+The `RegisterSerialize` class does the same thing but added the password field with the modality of ``write_only`. 
+
+While the `LoginSerializer`class does the trick of the authentication. The method called `validate` requires the data as the argument which passes it to the `authenticate` method from the rest framework. 
+
+By doing so, the user will be verified with their token that is already inside the data passed. If the token provided matches, the user will be logged in. But if otherwise,  a validation error will be sent.
+
+Furthermore, the views for the registration and login shall be done inside the file named `api.py` in our `account` app. 
+
+```python
+from django.contrib import auth
+from rest_framework import generics, permissions, serializers
+from rest_framework.response import Response
+from knox.models import AuthToken
+from .serializers import UserSerializer, RegisterSerializer, LoginSerializer
+
+class SignUpAPI(generics.GenericAPIView):
+    serializer_class = RegisterSerializer
+
+    def post(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        user = serializer.save()
+        token = AuthToken.objects.create(user)
+        return Response({
+            "users": UserSerializer(user, context=self.get_serializer_context()).data,
+            "token": token[1]
+        })
+
+
+class SignInAPI(generics.GenericAPIView):
+    serializer_class = LoginSerializer
+
+    def post(self, request):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        user = serializer.validated_data
+        return Response({
+            "user": UserSerializer(user, context=self.get_serializer_context()).data,
+            "token": AuthToken.objects.create(user)[1]
+        })
+
+
+class MainUser(generics.RetrieveAPIView):
+  permission_classes = [
+      permissions.IsAuthenticated
+  ]
+  serializer_class = UserSerializer
+
+  def get_object(self):
+    return self.request.user
+```
+
+Here are the explanations for each of the classes.
+
+As for the `RegisterAPI` and `LoginAPI` classes, we are providing the serializers made earlier to the `serializer_class` due to the class-based views that we used.
+
+The method `post` will be triggered whenever a user is try to register by sending data and a token is generated at the same time. For every user that registered successfully, a token will be sent along with the data for login.
+
```

Comment:
Change to " The method `post` will be triggered whenever a user tries to register by sending data and a token is generated simultaneously. For every user that registered successfully, a token will be sent along with the data for login."